pkg/eth/retriever_test.go

```diff
@@ -0,0 +1,344 @@
+			payload := mocks.MockConvertedPayload
+			payload.Block = newMockBlock(1010101)
+			err := repo.Publish(payload)
+			Expect(err).ToNot(HaveOccurred())
+			num, err := retriever.RetrieveFirstBlockNumber()
+			Expect(err).ToNot(HaveOccurred())
+			Expect(num).To(Equal(int64(1010101)))
+		})
+
+		It("Gets the number of the first block that has data in the database", func() {
+			payload1 := mocks.MockConvertedPayload
+			payload1.Block = newMockBlock(1010101)
+			payload2 := payload1
+			payload2.Block = newMockBlock(5)
+			err := repo.Publish(payload1)
+			Expect(err).ToNot(HaveOccurred())
+			err = repo.Publish(payload2)
+			Expect(err).ToNot(HaveOccurred())
+			num, err := retriever.RetrieveFirstBlockNumber()
+			Expect(err).ToNot(HaveOccurred())
+			Expect(num).To(Equal(int64(5)))
+		})
+	})
+
+	Describe("RetrieveLastBlockNumber", func() {
+		It("Throws an error if there are no blocks in the database", func() {
+			_, err := retriever.RetrieveLastBlockNumber()
+			Expect(err).To(HaveOccurred())
+		})
+		It("Gets the number of the latest block that has data in the database", func() {
+			err := repo.Publish(mocks.MockConvertedPayload)
+			Expect(err).ToNot(HaveOccurred())
+			num, err := retriever.RetrieveLastBlockNumber()
+			Expect(err).ToNot(HaveOccurred())
+			Expect(num).To(Equal(int64(1)))
+		})
+
+		It("Gets the number of the latest block that has data in the database", func() {
+			payload := mocks.MockConvertedPayload
+			payload.Block = newMockBlock(1010101)
+			err := repo.Publish(payload)
+			Expect(err).ToNot(HaveOccurred())
+			num, err := retriever.RetrieveLastBlockNumber()
+			Expect(err).ToNot(HaveOccurred())
+			Expect(num).To(Equal(int64(1010101)))
+		})
+
+		It("Gets the number of the latest block that has data in the database", func() {
+			payload1 := mocks.MockConvertedPayload
+			payload1.Block = newMockBlock(1010101)
+			payload2 := payload1
+			payload2.Block = newMockBlock(5)
+			err := repo.Publish(payload1)
+			Expect(err).ToNot(HaveOccurred())
+			err = repo.Publish(payload2)
+			Expect(err).ToNot(HaveOccurred())
+			num, err := retriever.RetrieveLastBlockNumber()
+			Expect(err).ToNot(HaveOccurred())
+			Expect(num).To(Equal(int64(1010101)))
+		})
+	})
+
+	Describe("RetrieveGapsInData", func() {
+		It("Doesn't return gaps if there are none", func() {
+			payload0 := mocks.MockConvertedPayload
+			payload0.Block = newMockBlock(0)
+			payload1 := mocks.MockConvertedPayload
+			payload2 := payload1
+			payload2.Block = newMockBlock(2)
+			payload3 := payload2
+			payload3.Block = newMockBlock(3)
+			err := repo.Publish(payload0)
+			Expect(err).ToNot(HaveOccurred())
+			err = repo.Publish(payload1)
+			Expect(err).ToNot(HaveOccurred())
+			err = repo.Publish(payload2)
+			Expect(err).ToNot(HaveOccurred())
+			err = repo.Publish(payload3)
+			Expect(err).ToNot(HaveOccurred())
+			gaps, err := retriever.RetrieveGapsInData(1)
+			Expect(err).ToNot(HaveOccurred())
+			Expect(len(gaps)).To(Equal(0))
```

Comment:
shouldn't it find gap (block 1) between 0 and 2?

Reply:
[This](https://github.com/vulcanize/ipfs-blockchain-watcher/blob/eth-ipld-indexer/pkg/eth/retriever_test.go#L129) payload has block number of 1, so there is no gap. These retriever tests are sort of messy/confusing with all the different `payload` variables, will try to clean this up a bit.